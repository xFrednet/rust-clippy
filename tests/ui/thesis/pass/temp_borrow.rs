//@rustc-env: CLIPPY_PETS_PRINT=1
//@rustc-env: CLIPPY_STATS_PRINT=1
//@rustc-env: CLIPPY_PRINT_MIR=1

#![warn(clippy::borrow_pats)]

fn temp_borrow_arg_1(owned: String) {
    owned.len();
    owned.is_empty();
}

fn temp_borrow_mut_arg_1(mut owned: String) {
    owned.pop();
    owned.push('x');
    owned.clear();
}

fn temp_borrow_mixed_arg_1(mut owned: String) {
    owned.is_empty();
    owned.clear();
    owned.len();
}

fn temp_borrow_local_1() {
    let owned = String::new();
    owned.len();
    owned.is_empty();
}

fn temp_borrow_mut_local_1() {
    let mut owned = String::new();
    owned.pop();
    owned.push('x');
    owned.clear();
}

fn temp_borrow_mixed_local_1() {
    let mut owned = String::new();
    owned.is_empty();
    owned.clear();
    owned.len();
}

fn temp_borrow_mixed_mutltiple_1(a: String, mut b: String) {
    let c = String::new();
    let mut d = String::new();

    // TempBorrow
    a.is_empty();
    b.is_empty();
    c.is_empty();
    d.is_empty();

    // TempBorrowMut
    b.clear();
    d.clear();

    // TempBorrow
    a.is_empty();
    b.is_empty();
    c.is_empty();
    d.is_empty();
}

fn main() {}
