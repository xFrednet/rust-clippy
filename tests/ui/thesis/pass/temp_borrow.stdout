# "temp_borrow_arg_1"
- owned          : (Immutable, Owned    , Argument, NonCopy, PartDrop, UserDef  ) {Borrow, ArgBorrow}
- Body: Output(Unit)     , NotConst , Safe  , Sync , Free       {HasTempBorrow}
- Stats: BodyStats {
    return_relations_signature: 0,
    return_relations_found: 0,
    arg_relations_signature: 0,
    arg_relations_found: 0,
    arg_relation_possibly_missed_due_generics: 0,
    arg_relation_possibly_missed_due_to_late_bounds: 0,
    owned: OwnedStats {
        arg_borrow_count: 2,
        arg_borrow_mut_count: 0,
        arg_borrow_extended_count: 0,
        arg_borrow_mut_extended_count: 0,
        named_borrow_count: 0,
        named_borrow_mut_count: 0,
        borrowed_for_closure_count: 0,
        borrowed_mut_for_closure_count: 0,
        two_phased_borrows: 0,
    },
}

# "temp_borrow_mut_arg_1"
- owned          : (Mutable  , Owned    , Argument, NonCopy, PartDrop, UserDef  ) {Borrow, ArgBorrowMut}
- Body: Output(Unit)     , NotConst , Safe  , Sync , Free       {HasTempBorrowMut}
- Stats: BodyStats {
    return_relations_signature: 0,
    return_relations_found: 0,
    arg_relations_signature: 0,
    arg_relations_found: 0,
    arg_relation_possibly_missed_due_generics: 0,
    arg_relation_possibly_missed_due_to_late_bounds: 0,
    owned: OwnedStats {
        arg_borrow_count: 0,
        arg_borrow_mut_count: 3,
        arg_borrow_extended_count: 0,
        arg_borrow_mut_extended_count: 0,
        named_borrow_count: 0,
        named_borrow_mut_count: 0,
        borrowed_for_closure_count: 0,
        borrowed_mut_for_closure_count: 0,
        two_phased_borrows: 0,
    },
}

# "temp_borrow_mixed_arg_1"
- owned          : (Mutable  , Owned    , Argument, NonCopy, PartDrop, UserDef  ) {Borrow, ArgBorrow, ArgBorrowMut}
- Body: Output(Unit)     , NotConst , Safe  , Sync , Free       {HasTempBorrow, HasTempBorrowMut}
- Stats: BodyStats {
    return_relations_signature: 0,
    return_relations_found: 0,
    arg_relations_signature: 0,
    arg_relations_found: 0,
    arg_relation_possibly_missed_due_generics: 0,
    arg_relation_possibly_missed_due_to_late_bounds: 0,
    owned: OwnedStats {
        arg_borrow_count: 2,
        arg_borrow_mut_count: 1,
        arg_borrow_extended_count: 0,
        arg_borrow_mut_extended_count: 0,
        named_borrow_count: 0,
        named_borrow_mut_count: 0,
        borrowed_for_closure_count: 0,
        borrowed_mut_for_closure_count: 0,
        two_phased_borrows: 0,
    },
}

# "temp_borrow_local_1"
- owned          : (Immutable, Owned    , Local   , NonCopy, PartDrop, UserDef  ) {Borrow, ArgBorrow}
- Body: Output(Unit)     , NotConst , Safe  , Sync , Free       {NoArguments, HasTempBorrow}
- Stats: BodyStats {
    return_relations_signature: 0,
    return_relations_found: 0,
    arg_relations_signature: 0,
    arg_relations_found: 0,
    arg_relation_possibly_missed_due_generics: 0,
    arg_relation_possibly_missed_due_to_late_bounds: 0,
    owned: OwnedStats {
        arg_borrow_count: 2,
        arg_borrow_mut_count: 0,
        arg_borrow_extended_count: 0,
        arg_borrow_mut_extended_count: 0,
        named_borrow_count: 0,
        named_borrow_mut_count: 0,
        borrowed_for_closure_count: 0,
        borrowed_mut_for_closure_count: 0,
        two_phased_borrows: 0,
    },
}

# "temp_borrow_mut_local_1"
- owned          : (Mutable  , Owned    , Local   , NonCopy, PartDrop, UserDef  ) {Borrow, ArgBorrowMut}
- Body: Output(Unit)     , NotConst , Safe  , Sync , Free       {NoArguments, HasTempBorrowMut}
- Stats: BodyStats {
    return_relations_signature: 0,
    return_relations_found: 0,
    arg_relations_signature: 0,
    arg_relations_found: 0,
    arg_relation_possibly_missed_due_generics: 0,
    arg_relation_possibly_missed_due_to_late_bounds: 0,
    owned: OwnedStats {
        arg_borrow_count: 0,
        arg_borrow_mut_count: 3,
        arg_borrow_extended_count: 0,
        arg_borrow_mut_extended_count: 0,
        named_borrow_count: 0,
        named_borrow_mut_count: 0,
        borrowed_for_closure_count: 0,
        borrowed_mut_for_closure_count: 0,
        two_phased_borrows: 0,
    },
}

# "temp_borrow_mixed_local_1"
- owned          : (Mutable  , Owned    , Local   , NonCopy, PartDrop, UserDef  ) {Borrow, ArgBorrow, ArgBorrowMut}
- Body: Output(Unit)     , NotConst , Safe  , Sync , Free       {NoArguments, HasTempBorrow, HasTempBorrowMut}
- Stats: BodyStats {
    return_relations_signature: 0,
    return_relations_found: 0,
    arg_relations_signature: 0,
    arg_relations_found: 0,
    arg_relation_possibly_missed_due_generics: 0,
    arg_relation_possibly_missed_due_to_late_bounds: 0,
    owned: OwnedStats {
        arg_borrow_count: 2,
        arg_borrow_mut_count: 1,
        arg_borrow_extended_count: 0,
        arg_borrow_mut_extended_count: 0,
        named_borrow_count: 0,
        named_borrow_mut_count: 0,
        borrowed_for_closure_count: 0,
        borrowed_mut_for_closure_count: 0,
        two_phased_borrows: 0,
    },
}

# "temp_borrow_mixed_mutltiple_1"
- a              : (Immutable, Owned    , Argument, NonCopy, PartDrop, UserDef  ) {Borrow, ArgBorrow}
- b              : (Mutable  , Owned    , Argument, NonCopy, PartDrop, UserDef  ) {Borrow, ArgBorrow, ArgBorrowMut}
- c              : (Immutable, Owned    , Local   , NonCopy, PartDrop, UserDef  ) {Borrow, ArgBorrow}
- d              : (Mutable  , Owned    , Local   , NonCopy, PartDrop, UserDef  ) {Borrow, ArgBorrow, ArgBorrowMut}
- Body: Output(Unit)     , NotConst , Safe  , Sync , Free       {HasTempBorrow, HasTempBorrowMut}
- Stats: BodyStats {
    return_relations_signature: 0,
    return_relations_found: 0,
    arg_relations_signature: 0,
    arg_relations_found: 0,
    arg_relation_possibly_missed_due_generics: 0,
    arg_relation_possibly_missed_due_to_late_bounds: 0,
    owned: OwnedStats {
        arg_borrow_count: 8,
        arg_borrow_mut_count: 2,
        arg_borrow_extended_count: 0,
        arg_borrow_mut_extended_count: 0,
        named_borrow_count: 0,
        named_borrow_mut_count: 0,
        borrowed_for_closure_count: 0,
        borrowed_mut_for_closure_count: 0,
        two_phased_borrows: 0,
    },
}

# "main"
- Body: Output(Unit)     , NotConst , Safe  , Sync , Free       {NoArguments}
- Stats: BodyStats {
    return_relations_signature: 0,
    return_relations_found: 0,
    arg_relations_signature: 0,
    arg_relations_found: 0,
    arg_relation_possibly_missed_due_generics: 0,
    arg_relation_possibly_missed_due_to_late_bounds: 0,
    owned: OwnedStats {
        arg_borrow_count: 0,
        arg_borrow_mut_count: 0,
        arg_borrow_extended_count: 0,
        arg_borrow_mut_extended_count: 0,
        named_borrow_count: 0,
        named_borrow_mut_count: 0,
        borrowed_for_closure_count: 0,
        borrowed_mut_for_closure_count: 0,
        two_phased_borrows: 0,
    },
}

